README.md
# Timerizer

A simple Ruby time helper, just like the Rails ones. You know, without the Rails.

```rb
5.minutes.ago
# => 2012-07-17 10:03:03 -0700
5.months.ago
# => 2012-02-17 10:08:03 -0800
1.minute.from_now
# => 2012-07-17 10:09:03 -0700
5.minutes.before(Time.new(2012, 1, 1, 3, 45, 00))
# => 2012-01-01 03:40:00 -0800
1.year.after(Time.new(2012, 1, 1, 3, 45, 00))
# => 2013-01-01 03:45:00 -0800
```

You can create durations `.seconds`, `.minutes`, etc., or with
`Timerizer::Duration`:

```rb
5.minutes == Timerizer::Duration.new(minutes: 5)
# => true
1.hour == Timerizer::Duration.new(hour: 1)
# => true
(1.week 1.day) == Timerizer::Duration.new(weeks: 1, days: 1)
# => true
(2.weeks) == Timerizer::Duration.new(days: 14)
# => true
```

Plus, it can convert these durations to strings:

```rb
Time.since(Date.yesterday).to_s
# => "1 day, 9 hours, 58 minutes, 3 seconds"
(1.year 3.minutes 4.seconds).to_s(:short)
# => "1yr 3min"
(5.hours 3.minutes).to_s(:micro)
# => "5h"
```

It also has this nice syntax for 'wall clock' times:

```rb
time = Timerizer::WallClock.new(5, 45, :pm)
Date.yesterday.at time
# => 2012-07-16 17:45:00 -0700
Date.tomorrow.at(WallClock.new(19, 30, 30))
# => 2012-07-18 19:30:30 -0700
```

## Requirements

Currently, Timerizer works in Ruby 2.0+

## Building and Installation

If you want to install Timerizer:

```
$ gem install timerizer
```

To build and install Timerizer yourself, do the following:

```
$ git clone git://github.com/kylewlacy/timerizer.git
$ cd timerizer
$ rake build
$ gem install ./pkg/timerizer-*.gem
```

## Documentation

Documentation can be found on [RubyDoc.info](http://rdoc.info/github/kylewlacy/timerizer/master/frames)


CHANGELOG.md
# Changelog

## [0.3.3] - 2019-09-10

### Changed

- **Timerizer now only supports Ruby 2.4+**. While earlier Ruby versions may still work, Ruby 2.4 is the oldest release that is still receiving security updates. So, you should update if you haven't already!
- [#10](https://github.com/kylewlacy/timerizer/pull/10) ([@muellerj](https://github.com/muellerj)): Fix warnings for Ruby 2.6 and update libraries

## [0.3.2] - 2018-12-28

### Fixed

- [#9](https://github.com/kylewlacy/timerizer/pull/9) ([@jdickey](https://github.com/jdickey)): Hotfix for issue preventing the library to load due to incorret gemspec configuration

## [0.3.1] - 2018-12-26

### Added

- [#8](https://github.com/kylewlacy/timerizer/pull/8) ([@jdickey](https://github.com/jdickey)): Added `Duration#to_rounded_s` to convert a duration to a human-readable string by rounding the duration first. This provides a more natural representation of a string.
    - The default format of `#to_rounded_s` is the new `:min_long` format, which is similar to the `:long` format except it excludes weeks.

## [0.3.0] - 2017-10-02

### Breaking changes

- **Timerizer now only supports Ruby 2.0+**!
- **`RelativeTime` is now `Timerizer:Duration`**. If you used the explicit
  `RelativeTime` constructor, you will need to update your code accordingly
  when updating to Timerizer 0.3.0+!
- **`WallClock` has moved into the `Timerizer` module**, so if you were using
  `WallClock`, all references will need to be updated to `Timerizer::WallClock`!
- The two-argument constructor of `RelativeTime` has been removed (i.e: `RelativeTime.new(5, :minutes)`). The equivalent with `Duration` now uses a hash: `Timerizer::Duration.new(minutes: 5)`
- `RelativeTime#average`, `RelativeTime#average!`, `RelativeTime#unaverage` and `RelativeTime#unaverage!` are no more. The equivalent methods are now `Duration#denormalize` (replacing `#unaverage`) and `Duration#normalize` (replacing `#average`), and both mutating versions are no more. **Note that `#average`/`#unaverage` do not give the same results as `#denormalize`/`#normalize`**, so be wary if you depended on the exact values.
- Remove `RelativeTime#{unit}` and `RelativeTime#in_{unit}` methods. These methods never really worked as intended (as seen in [issue #6](https://github.com/kylewlacy/timerizer/issues/6)).
  The closest equivalents are `Duration#to_unit` method (which takes a unit
  as a symbol), and the `Duration#to_{unit}` methods (`#to_seconds`,
  `#to_minutes`, etc.) . These methods live under a different name because the
  results differ from the former `#in_{unit}`/`#{unit}` methods.
- The `units`, `units_in_seconds` and `units_in_months` class methods
  from `RelativeTime` have been removed. See the new `Duration::UNITS` and
  `Duration::UNIT_ALIASES` constants for a replacement.
- Comparisons on `Duration` differ from the former comparisons on
  `RelativeTime`. Durations are normalized before comparing, so some equality
  tests may now return `true` (for example, `30.days == 1.month` now returns
  true due to normalization).

### Changed

- Timerizer now depends on Ruby 2.0+
- `RelativeTime` is now `Duration`
- `Duration` and `WallClock` have been moved into the new `Timerizer` module
- `Duration#new` has been reworked. It now takes a single hash mapping units
  to unit quantities. Unlike the old constructor, it takes arbitrary units
  now, not just `:seconds` and `:months`.
- `Duration#to_s` now uses `"mo"` as a shorthand instead of `"mn"`.
- `Duration#to_s` will now never return decades, centuries, or millennia in
  the default string formats. The changes to `Duration#to_s` now accept
  user-defined formatters, so clients can restore the old behavior with a
  custom formatter if needed.
- `Duration#to_s` now normalizes and denormalizes before printing. This means
  `30.days.to_s` will be equivalent to `1.month.to_s` by default.
- `Duration#==` now normalizes before comparison, so some comparions may be
  true that were false previously.

### Added

- Added `Duration#to_unit` to convert a duration to a specific unit. `#to_unit`
  normalizes the duration, so return values are more intuitive than the old
  conversion methods (for example, `1.month.to_unit(:seconds)` returns `30`).
- Added `Duration#to_units` to convert a duration to multiple significant units.
  This is similar to `Duration#to_unit`, except it can return multiple units
  at once. Example: `90.minutes.to_units(:hours, :minutes)` returns
  `{hours: 1, minutes: 30}`. See the docs for more details.
- Add `Duration#to_{unit}` helper methods (`#to_seconds`, `#to_minutes`, etc.).
- Added new `Duration#normalize` and `Duration#denormalize` to convert
  between second-based and month-based units.
- `Duration` now implements `#<=>` and `Comparable` to get all comparison
  operators (instead of just `==` and `!=`).
- `Duration` now supports representing negative durations of time.
- `Duration` now implements `#-@` (unary negation).
- `Duration` now implements `#*` (multiplication) and `#/` (divison) to multiply
  and divide by scalar values (currently only `Integer`s are supported).
- `Duration#to_s` now takes user-defined options. See the docs for more details.

### Fixed

- `Duration#to_s` no longer returns `""` (empty string) for empty durations
  (such as `0.seconds`). If _all_ units are empty, the returned string will
  be `"0 seconds"` (or equivalent based on formatting options).
- The new `Duration#to_{unit}` conversion methods fix conversions between
  second-based and month-based units by normalizing units first. This resolves
  [issue #6](https://github.com/kylewlacy/timerizer/issues/6).

### Removed

- `RelativeTime#average`, `RelativeTime#average!`, `RelativeTime#unaverage`, and
  `RelativeTime#unaverage!` have been removed. `Duration#denormalize` and
  `Duration#normalize` should now be used instead.
- `RelativeTime#in_{unit}` methods (`#in_seconds`, `#in_minutes`, etc.) have

spec/spec_helper.rb
require "bundler/setup"
require "timerizer"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  config.expose_dsl_globally = false

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/timerizer_spec.rb
require "spec_helper"

RSpec.describe Time do
  it "can be added or subtracted to Duration" do
    time = Time.new(2000, 1, 1, 3, 45, 00)
    expect(time + 5.minutes).to eq(Time.new(2000, 1, 1, 3, 50, 00))
    expect(time - 5.minutes).to eq(Time.new(2000, 1, 1, 3, 40, 00))
  end

  it "can be converted to a Date object" do
    time = Time.new(2000, 1, 1, 11, 59, 00)
    expect(time.to_date).to eq(Date.new(2000, 1, 1))
  end

  it "calculates the time between two Times" do
    time = 1.minute.ago
    expect(Time.until(1.minute.from_now).get(:seconds)).to be_within(1.0).of(60)
    expect(Time.since(1.hour.ago).get(:seconds)).to be_within(1.0).of(3600)

    expect(
      Time.between(1.minute.ago, 2.minutes.ago).get(:seconds)
    ).to be_within(1.0).of(60)

    expect(
      Time.between(Date.yesterday, Date.tomorrow).get(:seconds)
    ).to be_within(1.0).of(2 * 24 * 60 * 60)

    expect do
      Time.until(1.minute.ago)
    end.to raise_error(Time::TimeIsInThePastError)

    expect do
      Time.since(Date.tomorrow)
    end.to raise_error(Time::TimeIsInTheFutureError)
  end

  it "can be converted to a WallClock time" do
    time = Time.new(2000, 1, 1, 17, 58, 04)
    expect(time.to_wall).to eq(Timerizer::WallClock.new(5, 58, 04, :pm))
  end
end

RSpec.describe Date do
  it "can be converted to a Time object" do
    date = Date.new(2000, 1, 1)
    expect(date.to_time).to eq(Time.new(2000, 1, 1))
  end

  it "returns the number of days in a month" do
    expect(Date.new(2000, 1).days_in_month).to eq(31)
    expect(Date.new(2000, 2).days_in_month).to eq(29)
    expect(Date.new(2001, 2).days_in_month).to eq(28)
  end

  it "returns the date yesterday and tomorrow" do
    yesterday = 1.day.ago.to_date
    tomorrow = 1.day.from_now.to_date

    expect(Date.yesterday).to eq(yesterday)
    expect(Date.tomorrow).to eq(tomorrow)
  end

  it "returns the time on a given date" do
     date = Date.new(2000, 1, 1)
     time = Timerizer::WallClock.new(5, 00, :pm)
     expect(date.at(time)).to eq(Time.new(2000, 1, 1, 17))
  end
end

RSpec.describe Integer do
  it "creates `Duration`s" do
    expect(1.minute.get(:seconds)).to eq(60)
    expect(3.hours.get(:seconds)).to eq(10800)
    expect(5.days.get(:seconds)).to eq(432000)
    expect(4.years.get(:months)).to eq(48)

    duration = 1.second 2.minutes 3.hours 4.days 5.weeks 6.months 7.years
    expect(duration.get(:seconds)).to eq(3380521)
    expect(duration.get(:months)).to eq(90)
  end

  it "defines helpers for all units" do
    Timerizer::Duration::UNIT_ALIASES.each do |unit_name, _|
      duration = 1.send(unit_name)
      expect(duration).to eq(Timerizer::Duration.new(unit_name => 1))
    end
  end
end


lib/timerizer.rb
require "date"
require_relative "./timerizer/core"

# {Time} class monkey-patched with {Timerizer::Duration} support.
class Time
  # Represents an error where two times were expected to be in the future, but were in the past.
  # @see #until
  class TimeIsInThePastError < ArgumentError; end
  # Represents an error where two times were expected to be in the past, but were in the future.
  # @see #since
  class TimeIsInTheFutureError < ArgumentError; end

  class << self
    alias_method :classic_new, :new

    def new(*args)
      begin
        Time.classic_new(*args)
      rescue ArgumentError
        if args.empty?
          Time.new
        else
          Time.local(*args)
        end
      end
    end
  end

  # def initialize(*args)
  #   self.classic_new(args)
  #   # if args.count == 0

  #   # else
  #   # end
  # end

  alias_method :add, :+
  def +(time)
    if time.is_a?(Timerizer::Duration)
      time.after(self)
    else
      self.add(time)
    end
  end

  alias_method :subtract, :-
  def -(time)
    if time.is_a?(Timerizer::Duration)
      time.before(self)
    else
      self.subtract(time)
    end
  end

  # Calculates the time until a given time
  # @param [Time] time The time until now to calculate
  # @return [Duration] The time until the provided time
  # @raise[TimeIsInThePastException] The provided time is in the past
  # @example
  #   Time.until(Time.new(2012, 12, 25))
  #     => 13 weeks, 2 days, 6 hours, 31 minutes, 39 seconds
  # @see Time#since
  # @see Time#between
  def self.until(time)
    raise TimeIsInThePastError if Time.now > time.to_time

    Time.between(Time.now, time)
  end

  # Calculates the time since a given time
  # @param [Time] time The time to calculate since now
  # @return [Duration] The time since the provided time
  # @raise[TimeIsInTheFutureException] The provided time is in the future
  # @example
  #   Time.since(Time.new(2011, 10, 31))
  #     => 46 weeks, 5 days, 18 hours, 26 minutes, 10 seconds
  # @see Time#since
  # @see Time#between
  def self.since(time)
    raise TimeIsInTheFutureError if time.to_time > Time.now

    Time.between(Time.now, time)
  end

  # Calculate the amount of time between two times.
  # @param [Time] time1 The initial time
  # @param [Time] time2 The final time
  # @return [Duration] Calculated time between time1 and time2
  # @example
  #   Time.between(1.minute.ago, 1.hour.ago)
  #     => 59.minutes
  # @note The two times are interchangable; which comes first doesn't matter
  # @see Time#until
  # @see Time#since
  def self.between(time1, time2)
    time_between = (time2.to_time - time1.to_time).abs

    Timerizer::Duration.new(seconds: time_between.round)
  end

  undef_method :to_date
  # Convert {Time} to {Date}.
  # @return [Date] {Time} as {Date}
  # @example
  #   Time.new(2000, 1, 1, 12, 30).to_date
  #     => #<Date: 2000-01-01 ((2451545j,0s,0n),+0s,2299161j)>
  def to_date
    Date.new(self.year, self.month, self.day)
  end

  undef_method :to_time
  # Convert self to {Time}.
  # @see Date#to_time
  def to_time
    self
  end

  # Converts {Time} to {Timerizer::WallClock}
  # @return [Timerizer::WallClock] {Time} as {Timerizer::WallClock}
  # @example
  #   time = Time.now.to_wall
  #   Date.tomorrow.at(time)
  #     => 2000-1-2 13:13:27 -0800
  #     # "Same time tomorrow?"
  def to_wall
    Timerizer::WallClock.new(self.hour, self.min, self.sec)
  end
end

# {Date} class monkey-patched with {Timerizer::Duration} helpers.
class Date
  # Return the number of days in a given month.
  # @return [Integer] Number of days in the month of the {Date}.
  # @example
  #   Date.new(2000, 2).days_in_month
  #     => 29
  def days_in_month
    days_in_feb = (not self.leap?) ? 28 : 29
    number_of_days = [
      31,  days_in_feb,  31,  30,  31,  30,
      31,  31,           30,  31,  30,  31
    ]

    number_of_days.fetch(self.month - 1)
  end

  undef_method :to_date
  # Return self as {Date}.
  # @see Time#to_date
  def to_date
    self
  end

  # Apply a time to a date
  # @example yesterday at 5:00
  #   Date.yesterday.at(WallClock.new(5, 00, :pm))
  #     => 2000-1-1 17:00:00 -0800
  def at(time)
    time.to_wall.on(self)
  end

  # Return tomorrow as {Date}.
  # @see Date#yesterday
  def self.tomorrow
    1.day.from_now.to_date
  end

  # Return yesterday as {Date}.
  # @see Date#tomorrow
  def self.yesterday
    1.day.ago.to_date
  end
end

# Monkey-patched {Integer} class enabled to return {Timerizer::Duration}s.
# @example
#   5.minutes
#   # => 5 minutes
# @see Timerizer::Duration::UNITS
class Integer
  private

  # @!macro [attach] _define_duration_unit
  #   @method $1(other = nil)
  #
  #   Return a {Timerizer::Duration} with `self` of the given unit. This
  #   method is a helper that is equivalent to
  #   `Timerizer::Duration::new($1: self)`.
  #
  #   @param [Timerizer::Duration, nil] other Another duration to add to the
  #     resulting duration, if present. This argument allows "chaining" multiple
  #     durations together, to combine multiple units succiently.
  #
  #   @return [Timerizer::Duration] the quantity of the unit in the duration.
  #
  #   @see Timerizer::Duration#initialize
  #
  #   @example
  #     n.$1 == Timerizer::Duration.new($1: n)
  #     5.minutes == Timerizer::Duration.new(minutes: 5)
  #     (1.week 1.day) == 8.days # "Chaining" multiple units
  #     (n.$1 x.minutes) == (n.$1 + x.minutes)
  def self._define_duration_unit(unit)
    define_method(unit) do |other = nil|
      duration = Timerizer::Duration.new(unit => self)

      if other.nil?
        duration
      else
        duration + other
      end
    end
  end

  public

  self._define_duration_unit(:seconds)
  self._define_duration_unit(:minutes)
  self._define_duration_unit(:hours)
  self._define_duration_unit(:days)
  self._define_duration_unit(:weeks)
  self._define_duration_unit(:months)
  self._define_duration_unit(:years)
  self._define_duration_unit(:decades)
  self._define_duration_unit(:centuries)
  self._define_duration_unit(:millennia)
  self._define_duration_unit(:second)
  self._define_duration_unit(:minute)
  self._define_duration_unit(:hour)
  self._define_duration_unit(:day)
  self._define_duration_unit(:week)
  self._define_duration_unit(:month)
  self._define_duration_unit(:year)
  self._define_duration_unit(:decade)
  self._define_duration_unit(:century)
  self._define_duration_unit(:millennium)
end


lib/timerizer/relative_time.rb
# Represents a relative amount of time. For example, '`5 days`', '`4 years`', and '`5 years, 4 hours, 3 minutes, 2 seconds`' are all RelativeTimes.
class RelativeTime
  UNITS = {
    seconds: {seconds: 1},
    minutes: {seconds: 60},
    hours: {seconds: 60 * 60},
    days: {seconds: 24 * 60 * 60},
    weeks: {seconds: 7 * 24 * 60 * 60},
    months: {months: 1},
    years: {months: 12},
    decades: {months: 12 * 10},
    centuries: {months: 12 * 100},
    millennia: {months: 12 * 1000}
  }

  UNIT_ALIASES = UNITS.merge(
    second: UNITS[:seconds],
    minute: UNITS[:minutes],
    hour: UNITS[:hours],
    day: UNITS[:days],
    week: UNITS[:weeks],
    month: UNITS[:months],
    year: UNITS[:years],
    decade: UNITS[:decades],
    century: UNITS[:centuries],
    millennium: UNITS[:millennia]
  )

  NORMALIZATION_METHODS = {
    standard: {
      months: { seconds: 30 * 24 * 60 * 60 },
      years: { seconds: 365 * 24 * 60 * 60 }
    },
    minimum: {
      months: { seconds: 28 * 24 * 60 * 60 },
      years: { seconds: 365 * 24 * 60 * 60 }
    },
    maximum: {
      months: { seconds: 31 * 24 * 60 * 60 },
      years: { seconds: 366 * 24 * 60 * 60 }
    }
  }

  @@units = {
    second: :seconds,
    minute: :minutes,
    hour: :hours,
    day: :days,
    week: :weeks,
    month: :months,
    year: :years,
    decade: :decades,
    century: :centuries,
    millennium: :millennia
  }

  @@in_seconds = {
    second: 1,
    minute: 60,
    hour: 3600,
    day: 86400,
    week: 604800
  }

  @@in_months = {
    month: 1,
    year: 12,
    decade: 120,
    century: 1200,
    millennium: 12000
  }

  # Average amount of time in a given unit. Used internally within the {#average} and {#unaverage} methods.
  @@average_seconds = {
    month: 2629746,
    year: 31556952
  }

  # Default syntax formats that can be used with #to_s
  # @see #to_s
  SYNTAXES = {
    micro: {
      units: {
        seconds: 's',
        minutes: 'm',
        hours: 'h',
        days: 'd',
        weeks: 'w',
        months: 'mn',
        years: 'y',
      },
      separator: '',
      delimiter: ' ',
      count: 1
    },
    short: {
      units: {
        seconds: 'sec',
        minutes: 'min',
        hours: 'hr',
        days: 'd',
        weeks: 'wk',
        months: 'mn',
        years: 'yr',
        centuries: 'ct',
        millennia: 'ml'
      },
      separator: '',
      delimiter: ' ',
      count: 2
    },
    long: {
      units: {
        seconds: ['second', 'seconds'],
        minutes: ['minute', 'minutes'],
        hours: ['hour', 'hours'],
        days: ['day', 'days'],
        weeks: ['week', 'weeks'],
        months: ['month', 'months'],
        years: ['year', 'years'],
        centuries: ['century', 'centuries'],
        millennia: ['millenium', 'millennia'],
      }
    }
  }

  # All potential units. Key is the unit name, and the value is its plural form.
  def self.units
    @@units
  end

  # Unit values in seconds. If a unit is not present in this hash, it is assumed to be in the {@@in_months} hash.
  def self.units_in_seconds
    @@in_seconds
  end

  # Unit values in months. If a unit is not present in this hash, it is assumed to be in the {@@in_seconds} hash.
  def self.units_in_months
    @@in_months
  end

  # Initialize a new instance of RelativeTime.
  def initialize(units = {})
    @seconds = 0
    @months = 0

    units.each do |unit, n|
      unit_info = self.class.resolve_unit(unit)
      @seconds += n * unit_info.fetch(:seconds, 0)
      @months += n * unit_info.fetch(:months, 0)
    end
  end

  # Compares two RelativeTimes to determine if they are equal
  # @param [RelativeTime] time The RelativeTime to compare
  # @return [Boolean] True if both RelativeTimes are equal
  # @note Be weary of rounding; this method compares both RelativeTimes' base units
  def ==(time)
    if time.is_a?(RelativeTime)
      @seconds == time.get(:seconds) && @months == time.get(:months)
    else
      false
    end
  end

  # Return the number of base units in a RelativeTime.
  # @param [Symbol] unit The unit to return, either :seconds or :months
  # @return [Integer] The requested unit count
  # @raise [ArgumentError] Unit requested was not :seconds or :months
  def get(unit)
    if unit == :seconds
      @seconds
    elsif unit == :months
      @months
    else
      raise ArgumentError
    end
  end

  # Determines the time between RelativeTime and the given time.
  # @param [Time] time The initial time.
  # @return [Time] The difference between the current RelativeTime and the given time
  # @example 5 hours before January 1st, 2000 at noon
  #   5.minutes.before(Time.new(2000, 1, 1, 12, 00, 00))
  #     => 2000-01-01 11:55:00 -0800
  # @see #ago
  # @see #after
  # @see #from_now
  def before(time)
    (-self).after(time)
  end

  # Return the time between the RelativeTime and the current time.
  # @return [Time] The difference between the current RelativeTime and Time#now
  # @see #before
  def ago
    self.before(Time.now)
  end

  # Return the time after the given time according to the current RelativeTime.
  # @param [Time] time The starting time
  # @return [Time] The time after the current RelativeTime and the given time
  # @see #before
  def after(time)
    time = time.to_time

    prev_day = time.mday
    prev_month = time.month
    prev_year = time.year

    units = self.to_units(:years, :months, :days, :seconds)

    date_in_month = self.class.build_date(
      prev_year + units[:years],
      prev_month + units[:months],
      prev_day
    )
    date = date_in_month + units[:days]

    Time.new(
      date.year,
      date.month,
      date.day,
      time.hour,
      time.min,
      time.sec
    ) + units[:seconds]
  end

  # Return the time after the current time and the RelativeTime.
  # @return [Time] The time after the current time
  def from_now
    self.after(Time.now)
  end

  @@units.each do |unit, plural|
    in_method = "in_#{plural}"
    count_method = plural
    superior_unit = @@units.keys.index(unit) + 1

    if @@in_seconds.has_key?(unit)
      define_method(in_method) do
        self.normalize.get(:seconds) / @@in_seconds[unit]
      end
    elsif @@in_months.has_key?(unit)
      define_method(in_method) do
        self.denormalize.get(:months) / @@in_months[unit]
      end
    end

    in_superior = "in_#{@@units.values[superior_unit]}"
    count_superior = @@units.keys[superior_unit]

    define_method(count_method) do
      time = self.send(in_method)
      if @@units.length > superior_unit
        time -= self.send(in_superior).send(count_superior).send(in_method)
      end
      time
    end
  end

  def to_unit(unit)
    unit_details = self.class.resolve_unit(unit)

    if unit_details.has_key?(:seconds)
      seconds = self.normalize.get(:seconds)
      (seconds.to_f / unit_details.fetch(:seconds)).to_i
    elsif unit_details.has_key?(:months)
      months = self.denormalize.get(:months)
      (months.to_f / unit_details.fetch(:months)).to_i
    else
      raise "Unit should have key :seconds or :months"
    end
  end

  def to_units(*units)
    sorted_units = self.class.sort_units(units).reverse

    _, parts = sorted_units.reduce([self, {}]) do |(remainder, parts), unit|
      part = remainder.to_unit(unit)
      new_remainder = remainder - RelativeTime.new(unit => part)

      [new_remainder, parts.merge(unit => part)]
    end

    parts
  end

  def normalize(method: :standard)
    normalized_units = NORMALIZATION_METHODS.fetch(method).reverse_each

    initial = [0.seconds, self]
    result = normalized_units.reduce(initial) do |result, (unit, normal)|
      normalized, remainder = result

      seconds_per_unit = normal.fetch(:seconds)
      unit_part = remainder.send(:to_unit_part, unit)

      new_normalized = normalized + (unit_part * seconds_per_unit).seconds
      new_remainder = remainder - RelativeTime.new(unit => unit_part)
      [new_normalized, new_remainder]
    end

    normalized, remainder = result
    normalized + remainder
  end

  def denormalize(method: :standard)
    normalized_units = NORMALIZATION_METHODS.fetch(method).reverse_each

    denormalized = 0.seconds
    remainder = self

    initial = [0.seconds, self]
    result = normalized_units.reduce(initial) do |result, (unit, normal)|
      denormalized, remainder = result

      seconds_per_unit = normal.fetch(:seconds)
      remainder_seconds = remainder.get(:seconds)

      num_unit = (remainder_seconds.to_f / seconds_per_unit).to_i
      num_seconds_denormalized = num_unit * seconds_per_unit

      denormalized += RelativeTime.new(unit => num_unit)
      remainder -= num_seconds_denormalized.seconds

      [denormalized, remainder]
    end

    denormalized, remainder = result
    denormalized + remainder
  end

  def -@
    RelativeTime.new(seconds: -@seconds, months: -@months)
  end

  # Add two {RelativeTime}s together.
  # @raise ArgumentError Argument isn't a {RelativeTime}
  # @see #-
  def +(time)
    raise ArgumentError unless time.is_a?(RelativeTime)
    RelativeTime.new(
      seconds: @seconds + time.get(:seconds),
      months: @months + time.get(:months)
    )
  end

  # Find the difference between two {RelativeTime}s.
  # @raise ArgumentError Argument isn't a {RelativeTime}
  # @see #+
  def -(time)
    raise ArgumentError unless time.is_a?(RelativeTime)
    RelativeTime.new(
      seconds: @seconds - time.get(:seconds),
      months: @months - time.get(:months)
    )
  end

  # Converts {RelativeTime} to {WallClock}
  # @return [WallClock] {RelativeTime} as {WallClock}
  # @example
  #   (17.hours 30.minutes).to_wall
  #     # => 5:30:00 PM
  def to_wall
    raise WallClock::TimeOutOfBoundsError if @months > 0
    WallClock.new(second: @seconds)
  end

  # Convert {RelativeTime} to a human-readable format.
  def to_s(format = :long, options = nil)
    syntax =
      case format
      when Symbol
        SYNTAXES.fetch(format)
      when Hash
        format
      else
        raise ArgumentError, "Expected #{format.inspect} to be a Symbol or Hash"
      end

    syntax = syntax.merge(options || {})

    if syntax[:count].nil? || syntax[:count] == :all
      count = @@units.count
    else
      count = syntax[:count]
    end

    syntax_units = syntax.fetch(:units)
    units = self.to_units(*syntax_units.keys).select {|unit, n| n > 0}

    separator = syntax[:separator] || ' '
    delimiter = syntax[:delimiter] || ', '
    units.take(count).map do |unit, n|
      unit_label = syntax_units.fetch(unit)

      singular, plural =
        case unit_label
        when Array
          unit_label
        else
          [unit_label, unit_label]
        end

        unit_name =
          if n == 1
            singular
          else
            plural || singular
          end

        [n, unit_name].join(separator)
    end.join(syntax[:delimiter] || ', ')
  end

  private

  # This method is like `#to_unit`, except it does not perform normalization
  # first. Put another way, this method is essentially the same as `#to_unit`
  # except it does not normalize the value first. It is similar to `#get` except
  # that it can be used with non-primitive units as well.
  #
  # @example
  # (1.year 1.month 365.days).to_unit_part(:month)
  # # => 13
  # # Returns 13 because that is the number of months contained exactly within
  # # the sepcified `RelativeTime`. Since "days" cannot be translated to an
  # # exact number of months, they *are not* factored into the result at all.
  #
  # (25.months).to_unit_part(:year)
  # # => 2
  # # Returns 2 becasue that is the number of months contained exactly within
  # # the specified `RelativeTime`. Since "years" is essentially an alias
  # # for "12 months", months *are* factored into the result.
  def to_unit_part(unit)
    unit_details = self.class.resolve_unit(unit)

    if unit_details.has_key?(:seconds)
      seconds = self.get(:seconds)
      (seconds.to_f / unit_details.fetch(:seconds)).to_i
    elsif unit_details.has_key?(:months)
      months = self.get(:months)
      (months.to_f / unit_details.fetch(:months)).to_i
    else
      raise "Unit should have key :seconds or :months"
    end
  end

  def self.resolve_unit(unit)
    UNIT_ALIASES[unit] or raise ArgumentError, "Unknown unit: #{unit.inspect}"
  end

  def self.sort_units(units)
    units.sort_by do |unit|
      unit_info = self.resolve_unit(unit)
      [unit_info.fetch(:months, 0), unit_info.fetch(:seconds, 0)]
    end
  end

  def self.mod_div(x, divisor)
    modulo = x % divisor
    [modulo, (x - modulo).to_i / divisor]
  end

  def self.month_carry(month)
    month_offset, year_carry = self.mod_div(month - 1, 12)
    [month_offset + 1, year_carry]
  end

  # Create a date from a given year, month, and date. If the month is not in
  # the range 1..12, then the month will "wrap around", adjusting the given
  # year accordingly (so a year of 2017 and a month of 0 corresponds with
  # 12/2016, a year of 2017 and a month of 13 correpsonds with 1/2018, and so
  # on). If the given day is out of range of the given month, then the
  # date will be nudged back to the last day of the month.
  def self.build_date(year, month, day)
    new_month, year_carry = self.month_carry(month)
    new_year = year + year_carry

    if Date.valid_date?(new_year, new_month, day)
      Date.new(new_year, new_month, day)
    else
      Date.new(new_year, new_month, -1)
    end
  end
end
